registry/reference.go

```diff
@@ -164,13 +175,26 @@ func (r Reference) ValidateReference() error {
 	if r.Reference == "" {
 		return nil
 	}
-	if _, err := r.Digest(); err == nil {
+
+	return fmt.Errorf("%w: invalid non-tag, non-digest reference: %v", errdef.ErrInvalidReference, r.Reference)
```

Comment:
`ValidateReference()` validates the reference as an empty string or a digest or a tag with smart detection. That is, this function accepts all forms, not just empty strings.

Reply:
Based on that, `Validate()` should not take a `isTag` parameter as it will auto detect the reference type.

Reply:
I'm a bit confused on how you're proposing we deal with the bug found in that case, maybe you can help me understand:

Given an input of a form such as this for example: `registry.example.com:5000/hello-world:<valid-sha256>`

In ParseReference (the function in `registry/reference.go`):

1. First we break it up on the `/`, and get `registry` on the LHS and `path` on the RHS.
2. We look for `@` in `path`; not finding one, we deem that it cannot be valid form A or B.
3. We look for a `:` in path, and finding (at least) one, we deem that it must be a *tag*, valid form C.
4. Then we call `Validate()`, without an `isTag` as you've suggested.
5. Auto-detection kicks in, and finds that `<valid-sha256>` is indeed a valid *digest*, because the information about it having been deemed a *tag* earlier is by this stage lost.

Reply:
I think the following two things are mixed:

1. Parsing and validating a reference string like `registry.example.com:5000/hello-world:latest`.
2. Validating a `Reference` object.

For `#2`, an example usage is
```go
ref := registry.Reference{
    Registry:   "localhost:5000",
    Repository: "hello-world",
    Reference:  "latest",
}
err := ref.Validate()
check(err)
```
Unlike `#1`, there is no symbols like `@`, `:` in `#2`.

Reply:
In other words, `Validate()` is not meant to be invoked in `ParseReference` but be used in other scenarios like `#2`.

Reply:
Hmm, right now I get why I was confused.  But it seems that the only caller (in this package) that calls `Validate` was `ParseReference`.  I guess we're keeping it for callers outside the library.  I'll make this change and send a new PR.

---
I had written this but not posted, then received and read your response; the the following is just for the record, but you can ignore it.

> Take the example where the ParseReference is called with the following input:
> 
> `registry.example.com:5000/hello-world:sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855`
> 
> This function checks for `@`, and does not find one, it would hence rule out *digest* as a possibility.  Subsequently, it discovers a `:`, and immediately asserts that this can only be a *tag*, leaving the deep verification to the `Validate` call.  So, it has determined that this token can only be a `tag` -- **TOC**.
> 
> Next, it will call the proposed `Validate`, the one without the boolean `isTag` in the function prototype.  This function inspects what it has available to it, which is:
> 
> ```
> registry.Reference{
>     Registry:   "localhost:5000",
>     Repository: "hello-world",
>     Reference:  "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
> }
> ```
> 
> Auto-detection at this point can only make the following assertions: a) that this is not a *tag*, but a *digest*, and b) that the digest is valid -- **TOU**.
> 
> So either we have to perform complete validation within `ParseReference` while we still have access to `isTag`, or we need to forward it.